src/ros_can/src/node/node_ros_can.cpp

```diff
@@ -0,0 +1,214 @@
+  while (stat == canOK) {
+    // Convert the CAN message to a ROS message and publish it
+    canInterpreter(id, msg, dlc, flag, time);
+
+    stat = canRead(hnd, &id, &msg, &dlc, &flag, &time);
+  }
+}
+
+void RosCan::canInterpreter(long id, unsigned char msg[8], unsigned int dlc, unsigned int flag,
+                            unsigned long time) {
+  switch (id) {
+    case MASTER_STATUS:
+      switch (msg[0]) {
+        case 0x31:  // Current AS State
+        {
+          if (msg[1] == 3)  // If AS State == Driving
+            this->goSignal = 1;
+          else
+            this->goSignal = 0;
+          opStatusPublisher();
+          break;
+        }
+        case 0x32:  // Current AS Mission
+        {
+          this->asMission = msg[1];
+          opStatusPublisher();
+          break;
+        }
+        case 0x33:  // Left Wheel RPM
+        {
+          rlRPM = updatewheelRPM(msg);
+          rlRPMStatus = 1;
+
+          if (statusMsgCnt != 4)
+            syncRPM();
+          else if (syncMode == 0)
+            WheelRPMPublisher();
+
+          break;
+        }
+        default:
+          break;
+      }
+      break;
+
+    case MASTER_IMU:
+      imuPublisher(msg);
+      break;
+    case TEENSY_C1:
+      if (msg[0] == 0x11) {  // Rear Right RPM
+        rrRPM = updatewheelRPM(msg);
+        rrRPMStatus = 1;
+
+        if (statusMsgCnt != 4)
+          syncRPM();
+        else if (syncMode == 1)
+          WheelRPMPublisher();
+      }
+      break;
+    case STEERING_ID:
+      steeringAnglePublisher(msg[1], msg[2]);
+      break;
+
+    default:
+      break;
+  }
+}
+
+/**
+ * @brief Function to publish the Operational Status
+ */
+void RosCan::opStatusPublisher() {
+  auto message = custom_interfaces::msg::OperationalStatus();
+  message.go_signal = goSignal;
+  message.as_mission = asMission;
+  operationalStatus->publish(message);
+}
+
+/**
+ * @brief Function to publish wheel rpm
+ */
+void RosCan::WheelRPMPublisher() {
+  auto message = custom_interfaces::msg::WheelRPM();
+  message.rl_rpm = rlRPM;
+  message.rr_rpm = rrRPM;
+  wheelRPM->publish(message);
+  rrRPMStatus = 0;
+  rlRPMStatus = 0;
+}
+
+/**
+ * @brief Function to publish the imu values
+ */
+void RosCan::imuPublisher(unsigned char msg[8]) {
+  // Publish the imu values to a topic
+  auto message = sensor_msgs::msg::Imu();
+  message.linear_acceleration.x = (msg[2] << 8) | msg[1];
+  message.linear_acceleration.y = (msg[4] << 8) | msg[3];
+  message.angular_velocity.z = (msg[6] << 8) | msg[5];
+  imu->publish(message);
+}
+
+/**
+ * @brief Function to publish the steering angle
+ */
+void RosCan::steeringAnglePublisher(unsigned char angleLSB, unsigned char angleMSB) {
+  // Publish the steering angle to a topic
+  int angle = (angleMSB << 8) | angleLSB;
+  auto message = custom_interfaces::msg::SteeringAngle();
+  message.steering_angle = angle;
+  steeringAngle->publish(message);
+}
+
+void RosCan::syncRPM() {
```

Comment:
This seems like it is missing something, as depending on the intervals you will publish in left wheel or right wheel receival, which makes sense. However, in this function, you do not associate the intervals with the wheels you are receiving. Essentially, this function only works if the first message you receive is the Masters message, as you are deciding to publish on master receival in case the first interval is too big. But it could be the opposite (transition to statusMsgCnt 1 be made by receiving from C1), which would leave everything upside down. Also, you don't need step 4 (case 3)